components/fxa-client/src/scoped_keys.rs

```diff
@@ -86,152 +66,25 @@ impl ScopedKeysFlow {
-        // Uncompressed form (see SECG SEC1 section 2.3.3).
-        // First byte is 4, then 32 bytes for x, and 32 bytes for y.
-        assert_eq!(pub_key_bytes.len(), 1 + 32 + 32);
-        assert_eq!(pub_key_bytes[0], 0x04);
-        let x = Vec::from(&pub_key_bytes[1..33]);
-        let x = base64::encode_config(&x, base64::URL_SAFE_NO_PAD);
-        let y = Vec::from(&pub_key_bytes[33..]);
-        let y = base64::encode_config(&y, base64::URL_SAFE_NO_PAD);
-        Ok(json!({
-            "crv": "P-256",
-            "kty": "EC",
-            "x": x,
-            "y": y,
-        })
-        .to_string())
+    pub fn get_public_key_jwk(&self) -> Result<Jwk> {
+        Ok(jwcrypto::extract_pub_key_jwk(&self.key_pair)?)
     }
 
-    fn get_peer_public_from_jwk(jwk: Jwk) -> Result<Vec<u8>> {
-        let x = base64::decode_config(jwk.x, base64::URL_SAFE_NO_PAD)?;
-        let y = base64::decode_config(jwk.y, base64::URL_SAFE_NO_PAD)?;
-        if jwk.kty != "EC" {
-            return Err(ErrorKind::UnrecoverableServerError("Only EC keys are supported.").into());
-        }
-        if jwk.crv != "P-256" {
-            return Err(
-                ErrorKind::UnrecoverableServerError("Only P-256 curves are supported.").into(),
-            );
-        }
-        if x.len() != (256 / 8) {
-            return Err(ErrorKind::UnrecoverableServerError("X must be 32 bytes long.").into());
-        }
-        if y.len() != (256 / 8) {
-            return Err(ErrorKind::UnrecoverableServerError("Y must be 32 bytes long.").into());
-        }
-        let mut peer_pub_key: Vec<u8> = vec![0x04];
-        peer_pub_key.extend_from_slice(&x);
-        peer_pub_key.extend_from_slice(&y);
-        Ok(peer_pub_key)
+    pub fn decrypt_keys_jwe(self, jwe: &str) -> Result<String> {
+        let params = DecryptionParameters::ECDH_ES {
+            local_key_pair: self.key_pair,
+        };
+        Ok(jwcrypto::decrypt_jwe(jwe, params)?)
     }
 
     pub fn encrypt_keys_jwe(self, jwk: &str, data: &[u8]) -> Result<String> {
         let jwk: Jwk = serde_json::from_str(&jwk)?;
-        let peer_public_key = Self::get_peer_public_from_jwk(jwk)?;
-        let epk_jwk: Jwk = serde_json::from_str(&self.generate_keys_jwk()?)?;
-        let (private_key, _) = self.key_pair.split();
-
-        let content_key = private_key.agree(&agreement::ECDH_P256, &peer_public_key)?;
-        let apu = "";
-        let alg = "A256GCM";
-        let apv = "";
-        let secret = Self::get_secret_from_ikm(content_key, apu, apv, alg)?;
-        let sealing_key = aead::SealingKey::new(&aead::AES_256_GCM, &secret.as_ref())?;
-        let header = JweHeader {
-            alg: "ECDH-ES".to_string(),
-            enc: "A256GCM".to_string(),
-            epk: epk_jwk,
-            apu: None,
-            apv: None,
+        let params = EncryptionParameters::ECDH_ES {
+            enc: EncryptionAlgorithm::A256GCM,
+            peer_jwk: jwk,
+            local_key_pair: self.key_pair,
```

Comment:
This seems risky! `self.key_pair` here is designed to be the jwk that we send in to a scoped-keys oauth flow, but here we're using it as the ephemeral key for generating a JWE. I think *in practice* we would never find ourselves re-using this key but there's nothing in the way the code is set up to ensure that.

Please refactor so that we don't pass in a keypair here, and instead `jwcrypto::encrypt_to_jwe` generates a fresh ephemeral keypair internally.